### PR TITLE
Protodetect enip dns 4388 v5

### PR DESCRIPTION
**Pull request by catenacyber**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/4388

Describe changes:
- Improves probing parsers for nfs, enip and dns

so as to avoid protocol detection confusion with TCP splitting

Modifies #6030 with formatting fixes

### CODE REVIEW

**Comment by catenacyber** [comment]
Force-pushed to fix CIFuzz other variant by improving further enip (thought I was on a new branch)

**Comment by victorjulien** [comment]
@jasonish can you review this?

File: rust/src/dns/dns.rs [hunk 1]
```diff
@@ -701,23 +701,45 @@ impl DNSState {
...
+fn probe_header_validity(header : DNSHeader, rlen: usize) -> (bool, bool, bool) {
+    let opcode = ((header.flags >> 11) & 0xf) as u8;
+    if opcode >= 7 {
+        //unassigned opcode
+        return (false, false, false);
+    }
+    if header.additional_rr as usize > rlen {
+        //not enough data for additional records
+        return (false, false, false);
+    }
```
Comment (jasonish) [inline rust/src/dns/dns.rs:713]: We can end up here if `dns_parse_request` fails with incomplete, but `dns_parse_header` passes.  In the case of TCP could we just need more data here?

Comment (jasonish) [reply rust/src/dns/dns.rs:713]: Nevermind, still results in the probe returning ALPROTO_UNKNOWN, which should lead to the parser being called again when more data is available. Is that correct @victorjulien .

**Comment by jasonish** [comment]
My one concern is that this would eliminate the possibility to alert on something like a DNS request that uses a bad opcode.  But not sure where to draw the line such that we don't accidentally parse something as DNS which is semantically not DNS, or being able to detect broken DNS implementations, or someone doing goofy stuff. Same goes for ENIP.

File: rust/src/dns/dns.rs (hunk 1 above)
Comment (catenacyber) [reply rust/src/dns/dns.rs:713]: > the probe returning ALPROTO_UNKNOWN, which should lead to the parser being called again when more data is available

That seems correct to me

**Comment by catenacyber** [comment]
> My one concern is that this would eliminate the possibility to alert on something like a DNS request that uses a bad opcode. But not sure where to draw the line such that we don't accidentally parse something as DNS which is semantically not DNS, or being able to detect broken DNS implementations, or someone doing goofy stuff. Same goes for ENIP.

I guess that, in case of a first junk message, we need to resort on the other direction to get the correct protocol

cf https://github.com/OISF/suricata/pull/5759

File: rust/src/dns/dns.rs (hunk 1 above)
Comment (catenacyber) [reply rust/src/dns/dns.rs:713]: > We can end up here if `dns_parse_request` fails with incomplete, but `dns_parse_header` passes. In the case of TCP could we just need more data here?

It now looks to me like there was a bug in `rlen` value

**Comment by catenacyber** [comment]
Replaced by https://github.com/OISF/suricata/pull/6182

**Comment by jasonish** [comment]
> My one concern is that this would eliminate the possibility to alert on something like a DNS request that uses a bad opcode. But not sure where to draw the line such that we don't accidentally parse something as DNS which is semantically not DNS, or being able to detect broken DNS implementations, or someone doing goofy stuff. Same goes for ENIP.

@victorjulien Any thoughts on protocol accuracy vs detection of a protocol doing wrong things?

**Pull request closed without merging**